src/schedule/zephyr_dp_schedule.c

```diff
@@ -0,0 +1,378 @@
+		 0x8e, 0x4c, 0x2c, 0x95, 0xba, 0x8b, 0x15, 0x45);
+
+DECLARE_TR_CTX(dp_tr, SOF_UUID(dp_sched_uuid), LOG_LEVEL_INFO);
+
+struct scheduler_dp_data {
+	struct list_item tasks;		/* list of active dp tasks */
+	struct k_spinlock lock;		/* synchronization between cores */
+};
+
+struct task_dp_pdata {
+	k_tid_t thread_id;		/* zephyr thread ID */
+	k_thread_stack_t *p_stack;	/* pointer to thread stack */
+	uint32_t ticks_period;		/* period the task should be scheduled in LL ticks */
+	uint32_t ticks_to_trigger;	/* number of ticks the task should be triggered after */
+	struct k_sem sem;		/* semaphore for task scheduling */
+};
+
+/*
+ * there's only one instance of DP scheduler for all cores
+ * Keep pointer to it here
+ */
+static struct scheduler_dp_data *dp_sch;
+
+static inline k_spinlock_key_t scheduler_dp_lock(void)
+{
+	return k_spin_lock(&dp_sch->lock);
+}
+
+static inline void scheduler_dp_unlock(k_spinlock_key_t key)
+{
+	k_spin_unlock(&dp_sch->lock, key);
+}
+
+/*
+ * function called event LL tick, always on primary core
+ *
+ * TODO:
+ * the scheduler should here calculate deadlines of all task and tell Zephyr about them
+ * Currently there's an assumption that the task is always ready to run
+ *
+ * TODO: tasks with budgets
+ */
+void scheduler_dp_ll_tick(void)
+{
+	struct list_item *tlist;
+	struct task *curr_task;
+	struct task_dp_pdata *pdata;
+	k_spinlock_key_t lock_key;
+
+	if (!dp_sch)
+		return;
+
+	lock_key = scheduler_dp_lock();
+	list_for_item(tlist, &dp_sch->tasks) {
+		curr_task = container_of(tlist, struct task, list);
+		pdata = curr_task->priv_data;
+
+		if (pdata->ticks_to_trigger == 0) {
+			if (curr_task->state == SOF_TASK_STATE_QUEUED) {
+				/* set new trigger time, start the thread */
+				pdata->ticks_to_trigger = pdata->ticks_period;
+				curr_task->state = SOF_TASK_STATE_RUNNING;
+				k_sem_give(&pdata->sem);
+			}
+		} else {
+			if (curr_task->state == SOF_TASK_STATE_QUEUED ||
+			    curr_task->state == SOF_TASK_STATE_RUNNING)
+				/* decrease num of ticks to re-schedule */
+				pdata->ticks_to_trigger--;
+		}
+	}
+	scheduler_dp_unlock(lock_key);
+}
+
+static int scheduler_dp_task_cancel(void *data, struct task *task)
+{
+	(void)(data);
+	k_spinlock_key_t lock_key;
+
+	/* this is asyn cancel - mark the task as canceled and remove it from scheduling */
+	lock_key = scheduler_dp_lock();
+
+	task->state = SOF_TASK_STATE_CANCEL;
+	list_item_del(&task->list);
+
+	scheduler_dp_unlock(lock_key);
+
+	return 0;
+}
+
+static int scheduler_dp_task_complete(void *data, struct task *task)
+{
+	(void)(data);
+	k_spinlock_key_t lock_key;
+	struct task_dp_pdata *pdata = task->priv_data;
+
+	/* this is asyn cancel - mark the task as canceled and remove it from scheduling */
+	lock_key = scheduler_dp_lock();
+
+	task->state = SOF_TASK_STATE_COMPLETED;
+	list_item_del(&task->list);
+
+	scheduler_dp_unlock(lock_key);
+
+	/* let the thread run once again to call task_complete in a thread context */
+	k_sem_give(&pdata->sem);
```

Comment:
...if you do need `schedule_task_complete()` I'll have to ask why here you do wake up the task and in `.cancel()` you don't

Reply:
*  SOF_TASK_STATE_COMPLETED  - the task will be removed from scheduling,
     *				calling schedule_task will add the task to processing again
     *				task_complete() will be called
     *  SOF_TASK_STATE_CANCEL     - the task will be removed from scheduling,
     *				calling schedule_task will add the task to processing again
     *				task_complete() won't be called

calling complete will call task_complete() **in the context of task thread**

but looking at #7200 - maybe it should be simpified